### PR TITLE
W-16640678: Deadlock when parentContext completion depends on the completion of a child

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 1]
```diff
@@ -94,7 +96,9 @@ public AbstractEventContext(FlowExceptionHandler exceptionHandler, int depthLeve
                               Optional<CompletableFuture<Void>> externalCompletion) {
     this.depthLevel = depthLevel;
     this.externalCompletion = externalCompletion.orElse(null);
-    externalCompletion.ifPresent(completableFuture -> completableFuture.thenAccept((aVoid) -> tryTerminate()));
+    if (this.externalCompletion != null) {
+      this.externalCompletion.thenAccept((aVoid) -> tryTerminate());
+    }
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:101]: why this change? does this have a perf impact?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 2]
```diff
@@ -211,72 +203,87 @@ private synchronized void responseDone(Either<Throwable, CoreEvent> result) {
       signalConsumerSilently(onResponseConsumer);
     }
     onResponseConsumerList.clear();
+
     tryComplete();
   }
 
   protected void tryComplete() {
-    boolean allChildrenComplete;
+    if (areAllChildrenComplete() && state.compareAndSet(STATE_RESPONSE_RECEIVED, STATE_RESPONSE_PROCESSED)) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:211]: is this on purpose that the state won't change until `areAllChildrenComplete()` is true?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 3]
```diff
@@ -211,72 +203,87 @@ private synchronized void responseDone(Either<Throwable, CoreEvent> result) {
...
+        }
+      }
+
+      state.set(STATE_COMPLETE);
+      tryTerminate();
+    }
+  }
 
+  private boolean areAllChildrenComplete() {
+    boolean allChildrenComplete;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:235]: remove this var, return direnctly

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 2 above)
Comment (elrodro83) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:211]: also, checking the state would be slower than iterating the child context?

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 4]
```diff
@@ -211,72 +203,87 @@ private synchronized void responseDone(Either<Throwable, CoreEvent> result) {
...
-        } finally {
-          parent.getChildContextsWriteLock().unlock();
-        }
-      });
-
-      result = null;
-      responsePublisher = null;
-    }
+  private boolean canTerminate() {
+    return (externalCompletion == null || externalCompletion.isDone()) && this.state.get() == STATE_COMPLETE;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:286]: the caller is already checking the state

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 5]
```diff
@@ -386,7 +392,7 @@ public void accept(MonoSink<CoreEvent> sink) {
       if (isResponseDone()) {
         signalPublisherSink(sink);
       } else {
-        synchronized (AbstractEventContext.this) {
+        synchronized (this) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:395]: you're using the publisher now, is this on purpose?

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java [hunk 6]
```diff
@@ -84,6 +116,29 @@ public void neverFailsWrapper(@Optional Chain optionalProcessors,
     }
   }
 
+  public void asyncChainToObjectStore(@Expression(REQUIRED) Object latch,
+                                      Chain chain,
+                                      CompletionCallback<Void, Void> callback) {
+    final CountDownLatch countDownLatch = (CountDownLatch) latch;
+    final AtomicInteger counter = new AtomicInteger(0);
+    // scheduler.execute(() -> {
```
Comment (elrodro83) [inline tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java:124]: remove?

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 1 above)
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:101]: * easier to put breakpoints
*  in theory it should reduce GC impact by avoiding a lambda that's not adding much value. 
* seems to me like abusing the Optional API, which is often misused as a replacement for an `if`

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 2 above)
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:211]: the current code already tests the children before testing the state. However, the ordering is more about logic than performance. I'm using a CAS instruction so if test is positive, the value would atomically change, and that would be had to rollback

File: tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java (hunk 6 above)
Comment (marianogonzalez) [reply tests/test-extensions/heisenberg-extension/src/main/java/org/mule/test/heisenberg/extension/HeisenbergScopes.java:124]: i will delete this test case alltogether. Ignore

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java (hunk 5 above)
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:395]: nope. will fix

File: core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java [hunk 7]
```diff
@@ -94,7 +97,9 @@ public AbstractEventContext(FlowExceptionHandler exceptionHandler, int depthLeve
                               Optional<CompletableFuture<Void>> externalCompletion) {
     this.depthLevel = depthLevel;
     this.externalCompletion = externalCompletion.orElse(null);
-    externalCompletion.ifPresent(completableFuture -> completableFuture.thenAccept((aVoid) -> tryTerminate()));
+    if (this.externalCompletion != null) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java:100]: check for present instead of not null

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — DISMISSED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      childCompletedLatch.release();
+    });
+
+    childContext.onTerminated((e, t) -> childTerminatedLatch.countDown());
+
+    executorService.submit(() -> childContext.success());
+
+    await(testLatch);
+
+    check(5000, 500, () -> {
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java:87]: use the `RECEIVE_TIMEOUT` constant

File: core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      assertThat(parentContext.isTerminated(), is(true));
+      assertThat(childContext.isTerminated(), is(true));
+      return true;
+    });
+  }
+
+  private void await(CountDownLatch latch) {
+    try {
+      if (!latch.await(1000000000, SECONDS)) {
+        throw new AssertionError("Latch timed out");
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java:97]: the test will timeout before this: just use the no timeout version, or put a meaningful lower value (30 secs?)

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java (hunk 9 above)
Comment (marianogonzalez) [reply core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java:97]: this was this long for debugging purposes and forgot to change it.

File: core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    childContext.onTerminated((e, t) -> childTerminatedLatch.countDown());
+
+    executorService.submit(() -> childContext.success());
+
+    await(testLatch);
+
+    check(5000, 500, () -> {
+      assertThat(parentContext.isTerminated(), is(true));
+      assertThat(childContext.isTerminated(), is(true));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/event/AsyncEventContextCompletionTestCase.java:89]: add a string description of the assertion failure

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — DISMISSED** [review]

**Comment by marianogonzalez** [comment]
--validate

**Review by elrodro83 — APPROVED** [review]
